### PR TITLE
VACMS-9235 Create slack plugin for messages

### LINKED ISSUE #9235: Add slack notification as a messages notification option
## User Story or Problem Statement

As a a group of editors in need of notification of something in the CMS (example a new VA Form is created) I need for my team to be notified in slack.

## Acceptance Criteria
- [x] Update existing [slack module](https://www.drupal.org/project/slack) to latest release 1.3
- [x] ~Update existing slack notification in post_api to use the new slack method~ - No change needed
- [x] Add slack as message option in code.

## Implementation steps
<optional>

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [x] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [x] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [x] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by swirtSJW**
## Description

closes #9235

Soft blocked by https://github.com/department-of-veterans-affairs/va.gov-team/issues/43239

## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As admin (admins are the only ones that can edit these fields that are normally altered by migration)
1. New VA Form
   - [x] Visit https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/node/add/va_form
   - [x] Fill in all fields in Forms DB data section.
   - [x] Fill in all required fields
   - [x] Save the node.
   - [ ] visit https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/admin/reports/dblog?type%5B%5D=slack
   - [x] validate that there is an entry for the node you just saved.  It will not be successful in lower environments because they can not post to slack.
   
image

2. VA Form changed file name (url)
   - [x] Edit the node you just created.
   - [x] Change the value for "Link to form"
   
image

   - [x] save the node
   - [ ] visit https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/admin/reports/dblog?type%5B%5D=slack
   - [x] validate that there is an entry for the node you just saved.  It will not be successful in lower environments because they can not post to slack.
   
   
image

3. VA Form marked as deleted
    - [x] Edit the node you just saved.
    - [x] Put a check in the box for 'Deleted'
   
image

   - [x] save the node
   - [ ] visit https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/admin/reports/dblog?type%5B%5D=slack
   - [x] validate that there is an entry for the node you just saved.   
   
image

4. VA Form changed title
    - [ ] Edit the node you just saved.
    - [ ] Change the ~Page~ Form Title (NOT the same as the node title or the page title)
  
image

   - [ ] save the node
   - [ ] visit https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/admin/reports/dblog?type%5B%5D=slack
   - [ ] validate that there is an entry for the node you just saved. 
   
image

7. Then validate Acceptance Criteria from issue
- [x] Update existing [slack module](https://www.drupal.org/project/slack) to latest release 1.3
- [x] Add slack as message option in code.

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [x] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [x] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Comment by swirtSJW** [comment]
image

**Comment by swirtSJW** [comment]
image

**Comment by swirtSJW** [comment]
Here is a run of tests that my local CMS posted to Slack #cms-notifications.
https://dsva.slack.com/archives/CJT90C0UT/p1655931171810679

They should be going to the VA forms channel but so far our Slack bot can not post to that channel yet.

Request is here: https://dsva.slack.com/archives/CJT90C0UT/p1655467200139979

**Comment by swirtSJW** [comment]
Import is still failing on the new message templates.  The only thing the three failures have in common are that they were clones of the one that is importing fine.   I am guessing cloning may be incomplete and I will  recreate them by hand.

**Comment by swirtSJW** [comment]
Request for opening up the access on our slack app bot is here
https://github.com/department-of-veterans-affairs/va.gov-team/issues/43239

**Comment by swirtSJW** [comment]
I triggered a rebuild to run cypress.

The failure seems unrelated
```
1) Content Release
       The content release page should normally display no in-process releases:
     AssertionError: Timed out retrying after 10000ms: Expected to find content: 'Front end has not been built yet.' but never did.
      at Context.eval (https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/__cypress/tests?p=tests/cypress/integration/behavioral/content_release.feature:72999:28)
      at Context.resolveAndRunStepDefinition (https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/__cypress/tests?p=tests/cypress/integration/behavioral/content_release.feature:27189:29)
      at Context.eval (https://pr9265-8elkd3yodgnxnr3xxaqsviripvlbcfjc.ci.cms.va.gov/__cypress/tests?p=tests/cypress/integration/behavioral/content_release.feature:26510:35)
```

**Comment by swirtSJW** [comment]
Second failure on the content release cypress test.
image

File: docroot/modules/custom/va_gov_notifications/src/Service/NotificationsManager.php [hunk 1]
```diff
@@ -45,29 +45,42 @@ public function __construct(MessageNotifier $message_notifier, EntityTypeManager
...
    * @param array $values
    *   Array of key value pairs that will be passed to the template.
+   * @param string $notifier_name
+   *   The machine name of the notifier method (email, slack).
    *
    * @return bool
    *   Boolean value denoting success or failure of the notification.
    */
-  public function send($template_name, $user_id, array $values = []) : bool {
+  public function send($template_name, $recipient, array $values = [], $notifier_name = 'email') : bool {
```
Comment (tjheffner) [inline docroot/modules/custom/va_gov_notifications/src/Service/NotificationsManager.php:58]: love this update

**Review by tjheffner — CHANGES_REQUESTED** [review]
Ran through the QA steps, noticed two things. These might just be an issue with lower environment and not true blockers, but want to call it out just in case.

Messages generated for all the form actions:
Screen Shot 2022-06-27 at 9 20 23 AM

One thing I found is the QA steps note the message for changed form title (4th case) under "Page title" field under the forms db data, but the message is actually triggered on the "Form title" field under the forms.

Screen Shot 2022-06-27 at 9 21 07 AM

Screen Shot 2022-06-27 at 9 26 34 AM
 The double content entry is when i edited the field noted in QA steps, then the other form title field that triggered the webhooks. Also see the "form" error noting an`Illegal choice slack in type element.` This was me clicking the link to the dblog in the qa steps... slack isn't an option in that list. Maybe because lower environment? Not really a breaking issue, message_notify seems to collect the rest of the logs but wanted to raise it just in case.

File: docroot/modules/custom/va_gov_notifications/src/Service/NotificationsManager.php (hunk 1 above)
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_notifications/src/Service/NotificationsManager.php:58]: Glad you like this.  It felt dirty.  Any time you have a function that broadens its scope use multiple data types it is bad....  I needed shower... but at least the name is honest  :)

File: docroot/modules/custom/va_gov_workflow/src/EventSubscriber/EntityEventSubscriber.php [hunk 2]
```diff
@@ -139,9 +139,20 @@ private function removeArchiveOption(FormBaseAlterEvent $event) {
    */
   protected function flagVaFormChanges(EntityInterface $entity) {
     if ($entity->bundle() === 'va_form') {
-      $this->flagger->flagFieldChanged('field_va_form_title', 'changed_title', $entity, "The form title of this form changed from '@old' to '@new' in the Forms DB.");
-      $this->flagger->flagFieldChanged(['field_va_form_url', 'uri'], 'changed_filename', $entity, "The file name (URL) of this form changed from '@old' to '@new' in the Forms DB.");
-      $this->flagger->flagFieldChanged('field_va_form_deleted', 'deleted', $entity, "The form was marked as deleted in the Forms DB.");
+      if ($this->flagger->flagFieldChanged('field_va_form_title', 'changed_title', $entity, "The form title of this form changed from '@old' to '@new' in the Forms DB.")) {
```
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_workflow/src/EventSubscriber/EntityEventSubscriber.php:143]: Sorry, this was my fault.  Good code, bad QA step.  I fixed. the QA step.

[the remark this answers is not in the document]
Comment (tjheffner) [reply docroot/modules/custom/va_gov_workflow/src/EventSubscriber/EntityEventSubscriber.php:143]: all good, i wanted to make sure with 3 "title" fields on these forms. resolved!

**Review by tjheffner — APPROVED** [review]
LGTM!
Screen Shot 2022-06-28 at 7 40 55 AM
Screen Shot 2022-06-28 at 7 41 52 AM

The only thing I notice is that the list of messages found at either `/admin/content/message` or `/admin/content/messages` changes the logs when a field the message is referencing changes. I don't change the title of the form until the third or fourth node save, but it's changed in all messages, including the previously generated ones, when I view the list after the fact. That seems a little off to me, but definitely not your code causing it

**Comment by swirtSJW** [comment]
> The only thing I notice is that the list of messages found at either `/admin/content/message` or `/admin/content/messages` changes the logs when a field the message is referencing changes. I don't change the title of the form until the third or fourth node save, but it's changed in all messages, including the previously generated ones, when I view the list after the fact. That seems a little off to me, but definitely not your code causing it

That is an interesting observation, It seems the tokens are a live connection to the referenced node rather than snapshot from when the message was created.  That will be important to keep in mind if we surface changes in a dashboard in the future.

**Comment by swirtSJW** [comment]
image

**Pull request merged**